livy/src/main/java/org/apache/zeppelin/livy/BaseLivyInterprereter.java

```diff
@@ -148,7 +167,23 @@ private String extractStatementResult(String result) {
 
   @Override
   public void cancel(InterpreterContext context) {
-    //TODO(zjffdu). Use livy cancel api which is available in livy 0.3
+    if (livyVersion.isCancelSupported()) {
+      try {
+        if (paragraphId2StmtIdMapping.containsKey(context.getParagraphId())) {
+          cancelStatement(paragraphId2StmtIdMapping.get(context.getParagraphId()));
+        }
+      } catch (LivyException e) {
+        LOGGER.error("Fail to cancel statement " +
+            paragraphId2StmtIdMapping.get(context.getParagraphId()) + " for paragraph " +
+            context.getParagraphId(), e);
+      } finally {
+        if (paragraphId2StmtIdMapping.containsKey(context.getParagraphId())) {
+          paragraphId2StmtIdMapping.remove(context.getParagraphId());
```

Comment:
remove doesn't throw when the key is absence, so we should be good without contrainsKey check